### DISCUSSION TITLE
Questions on issues during Deployment

### DISCUSSION THREAD

**Discussion by Josverl** [OfficeDev/TACO#8] [Q&A] [answered]
Hi,

@peakers1976, I moved your question to the Q&A section 

I have had a couple of attempts at deploying the application with little success. It could be that I am missing something but I have followed the deployment guide step by step.

The deployment of the script appears to work and I do not get any failure output.

![image](https://user-images.githubusercontent.com/18456459/180723845-054e5c19-4f32-46b7-9fb9-e6b30adbfaed.png)

The first problem I encountered is the secrets list/get permissions do not get applied for the service account, I need to manually add the account, the only accounts that get added are the app service principal name and the admin account I run the script.

![image](https://user-images.githubusercontent.com/18456459/180724355-afda39fa-6645-4ad2-bbbc-b59661ebc5a4.png)

The second problem is when I import the solution into PowerApps. the documentation states that I should be able to click next and define the variables, this does not happen, it only gives me the option to import.

![image](https://user-images.githubusercontent.com/18456459/180724826-0e2faedd-e61b-4a92-9f85-df2762cde060.png)

![image](https://user-images.githubusercontent.com/18456459/180725145-713a634f-723a-47d3-8018-aa2c93fd5ba2.png)

I then have to manually go through and update the variables in the solution 

VAR - FunctionKey: value of the host function key
VAR - Tenant: tenantname, for example contoso
VAR - application id: Application (Client) ID of the App registration in Azure AD
VAR - FunctionApp: name of the function app
VAR - TeamsvoicemanagementSPSite: SharePoint Site name

additionally this variable 

![image](https://user-images.githubusercontent.com/18456459/180725494-59c1fafb-250e-44f4-aa85-5641bd11280b.png)

After I update the SharePoint site, I need to edit all of the Sharepoint list entries 

![image](https://user-images.githubusercontent.com/18456459/180725658-a1215498-3f68-4dd9-8d6b-4b3d6387dbca.png)

Finally, none of the Cloud Flows are enabled by default "Not sure if this is by design" if i use the import-CallQueue flow to import auto attendants as per below steps:

Automatically adding Auto Attendants
Perform the following steps to import the configuration of the call queues into the SharePoint Lists:

Browse to https://make.powerapps.com/
From the left menu select solutions
In the solutions overview click on the ... next to Auto Attendant and Call Queue Management and select the option Edit
Search for the item Import-CallQueue click on the ... and select Details followed by Details in a new tab
From the top menu bar select the option run
Wait until the flow has completed

 first, I notice that the Get Secret has the incorrect secret name.

![image](https://user-images.githubusercontent.com/18456459/180726500-4dcb8e1d-f2f6-4004-919c-15d99f1217c8.png)

 I need to update it with the correct result

![image](https://user-images.githubusercontent.com/18456459/180726641-02d45846-07eb-43e0-bfc5-5d2c38b15133.png)

The URI is incorrect for the POST method so I needed to update that.

Finally, it fails with the following error 

![image](https://user-images.githubusercontent.com/18456459/180727075-6879e31c-6bf3-4ff6-b77e-504523f79be0.png)

Any advice would be appreciated as it seems to be that some automation is not working as it should.

Thank you,
@peakers1976,

_Originally posted by @peakers1976 in https://github.com/OfficeDev/TACO/discussions/5#discussioncomment-3219769_

**Comment by JohanVeldhuis** [comment]
@peakers1976 and @Josverl 

Thanks for your feedback and questions.

It looks like that during the creation of the Azure Key vault the service account is not added for some reason. We will review the script and will update it to correct this issue. For now you can manually add the Azure Key vault service account and assign the get and list permissions.

We saw some weird behavior indeed in some cases when importing the solution which requires that you update the variables manually. This is a remark we will add to our manual. The reason why the flows are disabled by default is because in your case the variables were not updated as part of the import process. If this happens the flows will be disabled to prevent errors occuring when running the flows.

The secret name is a good catch we need to look into this as it should have an entry in the key vault normally.

Hope this clarifies your questions.

Johan

**Comment by peakers1976** [comment]
Hi Johan,

Thank you for your reply

Any idea about the failure of the flow with the importation of call queues
and auto attendants?

It does not appear the automatic functions work importing the existing CQs
and AAs.

**Reply by JohanVeldhuis** [reply]
@peakers1976 not without more details tbh, in our tests it worked fine so I do expect something going wrong.
If you can share more details that would be great, for example do we receive anything back when firing the Azure function or not

**Comment by peakers1976** [comment]
I will investigate further and get back to you.

Thank you

**Comment by peakers1976** [comment]
Hi Johan,

I just wanted to check a couple of things with you.

Firstly I found where I need to change the variable in the deploy.ps1 for
the service account to receive the get/list permissions on the key vault.

The flows will not work, I suspect its because of the functions not working

In the deploy.ps1 code should this line be commented out?

[image: image.png]

If I do run that warmup.ps1 script it fails

[image: image.png]

I just have a feeling the reason the flow fails is that the function apps
are not set up correctly when the script deploys.

[image: image.png]

[image: image.png]

I was wondering if you could take some screenshots to identify what may be
missing on the function apps side

**Reply by JohanVeldhuis** [reply]
@peakers1976  correct the deploy script was having an incorrect variable we added to our list to fix. Unfortunately, all screenshots are broken so it is hard to say what is wrong in your case

**Comment by peakers1976** [comment]
I have run the deployment at least 10 times and I am not sure everything in
the scripts applied correctly. The script runs without fault but none of
the functions work when triggered by the flows.

**Comment by JohanVeldhuis** [comment]
@peakers1976 I looked at the code again and it looks like the JSON file being used in prod did not contain name changes which clarifies why the Key vault names did not match the ones used in the different flows and the reason why the flow didn't return any output. Keep in mind that although the deployment might succeed it does not guarantee that the flows are working because the deployment script does not look at permissions etc. for the service account. I am waiting for 2 other team members to complete the review which will result in the deployment script as well as the JSON file to be updated which should fix some of the issues.

What you can do in the meantime to troubleshoot the function which is not working is do the following:

- got to the Azure function
- from the left menu select **authentication**
- temporarily switch of the authentication by pressing **edit**
- select **functions** from the left menu
- select the export-autoattendant or export-callqueue function
- in the left menu select **Code + test**
- wait for the function to load
- open up the log view on the bottom and make sure it is set to Filesystem logs, you might get a warning but accept it for now
- on the top of the page you will see this bar: 
<img width="445" alt="image" src="https://user-images.githubusercontent.com/13047574/182378157-0167d357-159d-49d0-8eb6-34eb691baee4.png">
- select the option **test/run** 
- in the new pane select **run**
- wait until the function completed
- validate the output of the body and log section

**Comment by peakers1976** [comment]
That's great thanks for the advice Johan.

Appreciate the quick response.

**Reply by JohanVeldhuis** [reply]
thanks for reporting the issues you experienced so we could take the corrective actions

**Comment by peakers1976** [comment]
No problem, looking forward to using the app

**Comment by StepanFuchs** [comment]
Hello,

I am not sure if I should create a new discussion channel :), however I am facing strange error as well.

I configured all as described and in my case I am facing issue directly on Azure function when I am importing call queues to share point list

![image](https://user-images.githubusercontent.com/110665168/183050395-7c5b4930-3158-4e1d-89d0-55b76dbd1612.png)


Seems like PowerShell module load is broken. I have checked all variables, secrets, modules directly in function file store and all is ok.

Have anybody experienced same issue please?

Thank you

**Reply by JohanVeldhuis** [reply]
@StepanFuchs in your case it looks like it is trying to perform a split on the call queue target which it cannot do. 
Below is the codesniplet which does that part.

So in your case it is able to get the timeout action because else the value would not be null.
After this step it will check which account type it is forwarded to and will select the correct value which is either a user id, group id or phonenumber. 

Looking at the code it looks like PowerShell thinks it is a different datatype which does not contain the split operation.

``` PowerShell  
 # Retrieving displayname for timeout action target
    If($cq.TimeoutActionTarget -ne $null)
    {
        If($(Test-IsGuid -StringGuid $cq.TimeoutActionTarget) -and $cq.TimeoutAction -ne "SharedVoicemail")
        {
            $TimeoutActionTarget = Get-CsOnlineuser $cq.TimeoutActionTarget
        }
        ElseIf($cq.TimeoutAction -eq "SharedVoicemail")
        {
            $authHeader = Get-AuthenticationToken
            $TimeoutActionTarget = $(Get-GroupObjectInfo -Token $authHeader -ObjectId $cq.TimeoutActionTarget.Id).displayName
        }    
        Else
        {
            $TimeoutActionTarget = $($cq.TimeoutActionTarget).split(":")[0]
        }
    }
```

Based on this I would like to ask you to perform the following to get some more details on why it is failing:

- Open PowerShell and connect to Teams via the Teams PowerShell module
- Run the following cmdlet

``` PowerShell

$queues = Get-CsCallQueue | Select Name
```

- Once completed run **$queues[0]** to get the output and check which Call Queue is the one at the top
- Once identified run
``` PowerShell
$cq = Get-CsCallQueue -Name <CallQueueName>
$cq.TimeoutActionTarget
````
- provide the output here (mask it to hide the PII data please)

Thanks,
Johan

**Reply by StepanFuchs** [reply]
Hello Johan,

Many thanks for support.

This is the name of first object 

Name              
----              
ACS-CQ-allocations 

Here is the output from 

$cq = Get-CsCallQueue -Name $queues[0].Name 

 
![image](https://user-images.githubusercontent.com/110665168/183067518-f6c4edf8-cbc6-4562-9d4c-4a1b2f504485.png)


However as you can see 

$cq.TimeoutActionTarget 

Is blank

$cq.TimeoutActionTarget.getType() 

PS C:\windows\system32> $cq.TimeoutActionTarget.getType()
You cannot call a method on a null-valued expression.
At line:1 char:1
+ $cq.TimeoutActionTarget.getType()
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull 


Thank you

Stepan

**Reply by JohanVeldhuis** [reply]
@StepanFuchs in that case we need to modify the export script to see if the output is $null or "" as it looks like the $null check isn't 100% full proof. Let me added as a bug so we can keep track of it an fix the issue.

**Reply by JohanVeldhuis** [reply]
@StepanFuchs so did some testing but in my case if the timeoutactiontarget is empty it simply skips the entry. Which means that it likely to fail on another CQ instead of the first one which I initially thought. Now ofcourse the question is which one ;-) Do you have any CQ configured which for example forward to a phonenumber? If so can you tell me in which format the number was added?

**Reply by StepanFuchs** [reply]
Hello Johan,

I have checked all forward options and forward to PSTN is done with standard format 

tel:+420111111111 for example 

or we have forward to object ID, which is standard guid like 

45a59455-gg45-4nn9-be7e-000000000000

Alternatively can I disable  somehow WEBSITE_RUN_FROM_PACKAGE?

I tried it for better troubleshooting, but it doesn't work looks like :) 

Many thanks

Stepan

**Reply by StepanFuchs** [reply]
Another comment,

I am running now single CQ import 

![image](https://user-images.githubusercontent.com/110665168/183077948-69ccf7c1-7682-492c-afd4-52bc511af67a.png)

Over this CQ there is object forward to GUID and issue is the same. 

When I run complex CQ import I can see similar issues but on different lines 

![image](https://user-images.githubusercontent.com/110665168/183078534-b04af73d-cf99-47ad-ad73-625a87e6610b.png)

So root cause probably the same 

Thank you

Stepan

**Reply by JohanVeldhuis** [reply]
Stepan,

Yes there is a way to change it however that will remove the current content completely. So if you want to disable the website_run_from_package you need to perform the following steps:

- go to the Azure function
- from the left menu select **Advanced Tools** in the **Development Tools** section
- in the main pane select **go**
- from the top menu select **Debug Console** and select **PowerShell** from the menu
- wait till everything is loaded
- in the "explorer" view select **site**
- click on the 3 dots next to **wwwroot** and select the option **Download** to download a zip with all the content
- wait for the download the be completed
- go back to the Azure Function page and selection **Configuration** in the **Settings** section
- search for the entry **WEBSITE_RUN_FROM_PACKAGE** 
- delete the entry completely
- after a few seconds/minutes go to **Functions** in the **Functions** section and you will see everything has been removed
- go back to the **Debug Console** window
- inside the wwwroot put the content from the zip into the directory
- go back to the Azure Function page and validate the functions are visible again

After this you should be able to make modifications to the functions.

Regards,
Johan

**Reply by JohanVeldhuis** [reply]
@StepanFuchs it is a different issue but maybe same cause. This one is for the overflow action but they are using a similar logic. Can you tell me which overflow actions is configured for this call queue?

**Reply by StepanFuchs** [reply]
Hello Johan,

There is following forward 

![image](https://user-images.githubusercontent.com/110665168/183087216-5e883dae-9eda-4cc6-85d9-0848286b98d5.png)

![image](https://user-images.githubusercontent.com/110665168/183087786-e722b93b-e6ca-42e4-9d4d-22b9e8d815e0.png)


but it is as I described in previous entry forward to object GUID

I think, that I am overlooking some peace of puzzle.

Thank you

Stepan

**Reply by JohanVeldhuis** [reply]
you are not ;-) so this one is for the test rgs correct?
So focussing on the OverFlowActiontarget one what it tries to do is validate if the OverflowActionTarget is a GUID because if that is the case we can lookup the object. In your case the OverFlowActiontarget is empty which results in the Test-IsGuid will fail because it is missing a value as you can see below:

<img width="894" alt="image" src="https://user-images.githubusercontent.com/13047574/183091766-68bedf8a-6f4b-459f-b2ae-519f76e36d05.png">

Can you do me a favor and try this:

``` PowerShell
$cq = get-cscallqueue -name <Test queue>

If($cq.OverflowActionTarget.Id -ne $null)
{
 Write-Host "Value not empty"
}
````

For the upper screenshot is it from the test call queue also?

**Reply by StepanFuchs** [reply]
Hello Johan

Yes output is from testing CQ 

![image](https://user-images.githubusercontent.com/110665168/183106892-80dca96f-1cc8-400b-af1d-928acc1eef7d.png)


When I run you code 

$cq = get-cscallqueue -name USDGF-CQ-TESTRG

If($cq.OverflowActionTarget.Id -ne $null)
{
 Write-Host "Value not empty"
}

it returns nothing so it is empty looks like :) 

Below is how I handle it in different internal azure function and it works :)


if($CallQueue.OverflowActionTarget -eq $null)
{
    $CallQueueOutputs.Add("OverflowActionTarget","")
    $CallQueueOutputs.Add("OverflowActionType",$CallQueue.OverflowAction.ToString())
}
else
{
    if($CallQueue.OverflowActionTarget.Type -eq "User")
    {
        $user = Get-CsOnlineUser -Identity $CallQueue.OverflowActionTarget.Id | select UserPrincipalName
        $CallQueueOutputs.Add("OverflowActionTarget",$user.UserPrincipalName)
    }
    else
    {
        $CallQueueOutputs.Add("OverflowActionTarget",$CallQueue.OverflowActionTarget.Id.Replace("tel:",""))
    }
   
    $CallQueueOutputs.Add("OverflowActionType",$CallQueue.OverflowActionTarget.Type.ToString())
}

Thank you Stepan

**Reply by StepanFuchs** [reply]
Hello Johan,

I used PowerShell code from azure function and I validated it row by row and it looks, that issue is caused by additional property $cq.TimeoutActionTarget.Id where your code is without ID

If I change code with $cq.OverflowActionTarget.Id it works as expected

![image](https://user-images.githubusercontent.com/110665168/183371233-1632ef93-f847-4de0-a6a7-60e1a633ea9c.png)

When I run $cq.TimeoutActionTarget I can see output like bellow so 

![image](https://user-images.githubusercontent.com/110665168/183372028-862801ac-524c-49bf-a92c-0533ac8219ec.png)

so your code doesn't count with our format of object, however my feeling was, that it is used globally the same.

Thank you

Stepan

**Reply by JohanVeldhuis** [reply]
@Stepan ok good catch so in this case if we would update it from this:

```PowerShell
  If($cq.TimeoutActionTarget -ne $null)

    {

        If($(Test-IsGuid -StringGuid $cq.TimeoutActionTarget) -and $cq.TimeoutAction -ne "SharedVoicemail")

        {

            $TimeoutActionTarget = Get-CsOnlineuser $cq.TimeoutActionTarget

        }

        ElseIf($cq.TimeoutAction -eq "SharedVoicemail")

        {

            $authHeader = Get-AuthenticationToken

            $TimeoutActionTarget = $(Get-GroupObjectInfo -Token $authHeader -ObjectId $cq.TimeoutActionTarget.Id).displayName

        }    

        Else

        {

            $TimeoutActionTarget = $($cq.TimeoutActionTarget).split(":")[0]

        }

    }
```

to this:
``` PowerShell
  If($cq.TimeoutActionTarget -ne $null)

    {

        If($(Test-IsGuid -StringGuid $cq.TimeoutActionTarget.Id) -and $cq.TimeoutAction -ne "SharedVoicemail")

        {

            $TimeoutActionTarget = Get-CsOnlineuser $cq.TimeoutActionTarget.Id

        }

        ElseIf($cq.TimeoutAction -eq "SharedVoicemail")

        {

            $authHeader = Get-AuthenticationToken

            $TimeoutActionTarget = $(Get-GroupObjectInfo -Token $authHeader -ObjectId $cq.TimeoutActionTarget.Id).displayName

        }    

        Else

        {

            $TimeoutActionTarget = $($cq.TimeoutActionTarget.Id).split(":")[1]

        }

    }
```

So I modified the code based on your suggestion adding the .id property which fixed the issue if call forwarding is set to a single user and the phone last Else which was missing also the .Id.

No let's look at the next issues with the call overflow because that one has the .id property.

Regards,
Johan

**Reply by StepanFuchs** [reply]
Hello Johan,

I was able to go further with another call queue however flow is then failing for JSON not compatible schema 

![image](https://user-images.githubusercontent.com/110665168/183598686-a3303909-b3d0-4c7e-b0fb-9671b0e069bf.png)

I fixed it but then it doesn't create an item in to the SharePoint probably due to different data inputs from source. 

I think if we want to reuse this app we need to validate whole data types.

**Reply by JohanVeldhuis** [reply]
@StepanFuchs looks like the JSON schema needs to be validated on the import-callqueue flow. This is probably caused by the fact by generating the JSON using a queue which had the overflow action. Based on this it added datatypes which might in cases no values are assigned causes errors. Think this one was overlooked as we did cleanup the auto attendant one.

Below is how the JSON should look like:

```JSON
{

    "type": "array",

    "items": {

        "type": "object",

        "properties": {

            "Name": {

                "type": "string"

            },

            "AgentAlertTime": {},

            "OverflowThreshold": {},

            "OverflowAction": {},

            "OverflowActionTarget": {},

            "OverflowSharedTextToSpeechPrompt": {},

            "OverflowSharedVoicemailAudioFilePromptFileName": {},

            "TimeoutThreshold": {},

            "TimeoutAction": { },

            "TimeoutActionTarget": {},

            "TimeoutSharedTextToSpeechPrompt": {},

            "TimeoutSharedVoicemailAudioFilePromptFileName": {},

            "UseDefaultWelcomeMusic": {},

            "WelcomeMusicFileName": {},

            "UseDefaultMusicOnHold": {},

            "MusicOnHoldFileName": {}

        },

        "required": [

            "Name"

        ]

    }

}
```

Regards,
Johan

**Reply by StepanFuchs** [reply]
Hello Johan,

I am reviving further scope of the project. I notice another complication, that power app filter is set to DisplayName, which is not valid for our purpose. I would need to change it to email instead

![image](https://user-images.githubusercontent.com/110665168/183675977-7cb69b23-d8cd-4133-9eca-f3c3c4d26700.png)

otherwise I am not able to join the data. 

Generally solution is well done and it would help us to replace missing BRAC over CQ and AA, which in skype is available :) 

I hope it help you to tune the project for other as well as I know, that MS officially plans similar kind of tool in the end of next year. 

However we would need to validate whole project and fix issue by issue. 

Thank you

Stepan

**Reply by JohanVeldhuis** [reply]
Hi Stepan,

First of all thanks for the help and testing you did. It has uncovered some issues which we didn't spot ourselves, so we are really happy with your help and support. I understand your requirements are a bit different and it would require a full review of the solution. Feel free to provide the new features or enhancement you would like to see and we can take them into consideration. The example you provided is for example a good candidate but would require some rework in our case

**Comment by peakers1976** [comment]
Hi Johan,

Just wondering if you had a chance to look at the logs from the
export-callqueue function and shed some light on the problem?

**Reply by JohanVeldhuis** [reply]
@peakers1976 maybe I missed it but I don't see any logs from you but it is likely the same issue as Stepan is seeing

**Comment by peakers1976** [comment]
Possibly, I attached a log file to the email maybe it didn't go through.

Any chance you work for Microsoft? I noticed someone with the same name as
yours in the address book on teams.

**Reply by JohanVeldhuis** [reply]
yes ;-) so feel free to send it directly

**Comment by JohanVeldhuis** [comment, marked as answer]
All I am going to mark this discussion as closed. We released a new version of the Azure Function App which should fix most issues which are reported, a small description about the changes can be found: [here](https://github.com/OfficeDev/TACO/discussions/17)

**Comment by peakers1976** [comment]
Hi Johan,

I followed the steps to disable the authentication and output the logs to
the filesystem

Please see the attached log, it does not complete

**Comment by peakers1976** [comment]
Nice 😊

I will try not to annoy you to much.

If you have some spare time one day next week maybe we can touch base to
screen share and I can show you the issue.

Thanks Johan
